In [8]:
# %pip install bm25s
# %pip install spacy
# %pip install -U 'spacy[cuda12x]'
# %pip install rouge_score

In [1]:
import functions as fct
import bm25s
import numpy as np
import json

resource module not available on Windows


In [2]:
train_path = 'SCOTUS/train.json'
dev_path = 'SCOTUS/dev.json'

Summarization with legal-pegasus

In [11]:
text = """On March 5, 2021, the Securities and Exchange Commission charged AT&T, Inc. with repeatedly violating Regulation FD, and three of its Investor Relations executives with aiding and abetting AT&T's violations, by selectively disclosing material nonpublic information to research analysts. According to the SEC's complaint, AT&T learned in March 2016 that a steeper-than-expected decline in its first quarter smartphone sales would cause AT&T's revenue to fall short of analysts' estimates for the quarter. The complaint alleges that to avoid falling short of the consensus revenue estimate for the third consecutive quarter, AT&T Investor Relations executives Christopher Womack, Michael Black, and Kent Evans made private, one-on-one phone calls to analysts at approximately 20 separate firms. On these calls, the AT&T executives allegedly disclosed AT&T's internal smartphone sales data and the impact of that data on internal revenue metrics, despite the fact that internal documents specifically informed Investor Relations personnel that AT&T's revenue and sales of smartphones were types of information generally considered "material" to AT&T investors, and therefore prohibited from selective disclosure under Regulation FD. The complaint further alleges that as a result of what they were told on these calls, the analysts substantially reduced their revenue forecasts, leading to the overall consensus revenue estimate falling to just below the level that AT&T ultimately reported to the public on April 26, 2016. The SEC's complaint, filed in federal district court in Manhattan, charges AT&T with violations of the disclosure provisions of Section 13(a) of the Securities Exchange Act of 1934 and Regulation FD thereunder, and charges Womack, Evans and Black with aiding and abetting these violations. The complaint seeks permanent injunctive relief and civil monetary penalties against each defendant. The SEC's investigation was conducted by George N. Stepaniuk, Thomas Peirce, and David Zetlin-Jones of the SEC's New York Regional Office. The SEC's litigation will be conducted by Alexander M. Vasilescu, Victor Suthammanont, and Mr. Zetlin-Jones. The case is being supervised by Sanjay Wadhwa."""

summary = fct.summarize(text, "legal-pegasus")

print(summary)

On March 5, 2021, the Securities and Exchange Commission charged AT&T, Inc. and three of its Investor Relations executives with aiding and abetting the company's violations of the antifraud provisions of Section 10(b) of the Securities Exchange Act of 1934 and Rule 10b-5 thereunder. According to the SEC's complaint, the company learned in March 2016 that a steeper-than-expected decline in its first quarter smartphone sales would cause its revenue to fall short of analysts' estimates for the quarter. The complaint alleges that to avoid falling short of the consensus revenue estimate for the third consecutive quarter, the executives made private, one-on-one phone calls to analysts at approximately 20 separate firms. On these calls, the SEC alleges that Christopher Womack, Michael Black, and Kent Evans told analysts that a steep decline in smartphone sales was a significant factor in the revenue shortfall. As a result of what they were told, the analysts substantially reduced their revenu

BM25 with custom_spacy

In [12]:
# Create a corpus for BM25
corpus = fct.sent_segmentation(text)

# Initialize and index BM25
retriever = bm25s.BM25(corpus=corpus)
retriever.index(bm25s.tokenize(corpus))  

# Process the summary
sum_sentences = fct.sent_segmentation(summary)

# Retrieve sentences
for sent in sum_sentences:
    results, scores = retriever.retrieve(bm25s.tokenize(sent), k=2)
    avg_score = np.mean(scores)
    print(sent + " ; " + str(avg_score))


Split strings:   0%|          | 0/10 [00:00<?, ?it/s]

DEBUG:bm25s:Building index from IDs objects


BM25S Count Tokens:   0%|          | 0/10 [00:00<?, ?it/s]

BM25S Compute Scores:   0%|          | 0/10 [00:00<?, ?it/s]

Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

On March 5, 2021, the Securities and Exchange Commission charged AT&T, Inc. and three of its Investor Relations executives with aiding and abetting the company's violations of the antifraud provisions of Section 10(b) of the Securities Exchange Act of 1934 and Rule 10b-5 thereunder. ; 8.678377


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

According to the SEC's complaint, the company learned in March 2016 that a steeper-than-expected decline in its first quarter smartphone sales would cause its revenue to fall short of analysts' estimates for the quarter. ; 8.918018


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

The complaint alleges that to avoid falling short of the consensus revenue estimate for the third consecutive quarter, the executives made private, one-on-one phone calls to analysts at approximately 20 separate firms. ; 8.376959


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

On these calls, the SEC alleges that Christopher Womack, Michael Black, and Kent Evans told analysts that a steep decline in smartphone sales was a significant factor in the revenue shortfall. ; 3.9654934


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

As a result of what they were told, the analysts substantially reduced their revenue forecasts, leading to the overall consensus Revenue Estimate falling to just below the level that AT&t ultimately reported to the public on April 26, 2016. ; 8.748906


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

The SEC seeks permanent injunctive relief and civil monetary penalties against each defendant. ; 5.397449


Split strings:   0%|          | 0/1 [00:00<?, ?it/s]

BM25S Retrieve:   0%|          | 0/1 [00:00<?, ?it/s]

The SEC investigation was conducted by George N. Stepaniuk, ; 2.639274


In [13]:
print(text)
print(summary)

On March 5, 2021, the Securities and Exchange Commission charged AT&T, Inc. with repeatedly violating Regulation FD, and three of its Investor Relations executives with aiding and abetting AT&T's violations, by selectively disclosing material nonpublic information to research analysts. According to the SEC's complaint, AT&T learned in March 2016 that a steeper-than-expected decline in its first quarter smartphone sales would cause AT&T's revenue to fall short of analysts' estimates for the quarter. The complaint alleges that to avoid falling short of the consensus revenue estimate for the third consecutive quarter, AT&T Investor Relations executives Christopher Womack, Michael Black, and Kent Evans made private, one-on-one phone calls to analysts at approximately 20 separate firms. On these calls, the AT&T executives allegedly disclosed AT&T's internal smartphone sales data and the impact of that data on internal revenue metrics, despite the fact that internal documents specifically in

In [14]:
print(len(text), len(summary))

2206 1312


In [15]:
score = fct.rouge_evaluations(text, summary)
print(score)

   Metric  Precision    Recall  F1-Score
0  rouge1   0.919811  0.560345  0.696429
1  rouge2   0.791469  0.481268  0.598566
2  rougeL   0.768868  0.468391  0.582143


# BERT

In [5]:
import nltk
from transformers import BertModel, BertTokenizer
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import numpy as np
import json

In [17]:
# Charger le modèle et le tokenizer BERT pré-entraîné
model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertModel.from_pretrained(model_name)

In [3]:
# Exemple de texte juridique
# document = """
# The law guarantees the fundamental rights and freedoms of every citizen.
# The judge rendered his verdict based on the 2023 law.
# The rights of citizens must be respected in all legal proceedings.
# The 1958 constitution is the basis of the current legal system.
# Every individual has the right to a fair trial under applicable laws.
# """

with open(train_path, 'r', encoding="utf-8") as f:
    train = json.load(f)

document = train[0]["raw_source"]

# with open('data_txt_save/train_0.txt', 'r', encoding="utf-8") as f:
#     document = f.read()

In [19]:
# Tokenisation en phrases
sentences = fct.sent_segmentation(document, method='nltk')

# Fonction pour obtenir les embeddings de phrases avec BERT
def get_sentence_embeddings(sentences, tokenizer, model):
    """Obtenir les embeddings de phrases avec BERT
    Args:
        sentences (List[str]): Liste des phrases à encoder
        tokenizer (transformers.PreTrainedTokenizer): Tokenizer BERT
        model (transformers.PreTrainedModel): Modèle BERT
    Returns:
        np.array: Tableau des embeddings de phrases    
    """
    embeddings = []
    for sentence in sentences:
        inputs = tokenizer(sentence, return_tensors='pt', padding=True, truncation=True)
        outputs = model(**inputs)
        embeddings.append(outputs.last_hidden_state.mean(dim=1).squeeze().detach().numpy())  # Moyenne des embeddings
    return np.array(embeddings)

# Obtenir les embeddings des phrases
sentence_embeddings = get_sentence_embeddings(sentences, tokenizer, model)

# K-Means

In [20]:
# Utiliser K-Means pour regrouper des phrases similaires
n_clusters = 5  # Nombre de clusters souhaités (peut être ajusté)
kmeans = KMeans(n_clusters=n_clusters, random_state=0)
kmeans.fit(sentence_embeddings)

# Étiquettes des clusters
labels = kmeans.labels_

In [21]:
cluster = {}
for i in range(n_clusters):
    print(f"\nCluster {i+1}:")
    cluster[i] = []  # Initialiser une liste pour chaque cluster
    for j, sentence in enumerate(sentences):
        if labels[j] == i:
            print(f"- {sentence}")
            # Ajouter 'sentence' au cluster 'i'
            cluster[i].append(sentence)
    # Afficher les clés du dictionnaire 'cluster'
print(cluster.keys())


Cluster 1:
- 01-1765.
- 01-1765.
- They do this or not do this?
- This element, this element, this
element?
- You
determine the facts.
- App.
- to Pet.
- for
Cert.
- H-15 to H-16 (citations omitted).
- See id., at 579, n. 10.
- App.
- to Pet.
- for Cert.
- H-15.
- to Pet.
- for
Cert.

Cluster 2:
- Awareness of this Court's
cases is not even required, so long as neither the reasoning nor
the result of the state-court decision contradicts them.
- Awareness of this Court's
cases is not even required, so long as neither the reasoning nor
the result of the state-court decision contradicts them.
- But they do
not have a right to not deliberate.
- They must deliberate and follow
the rules and laws as I state it to them.'"
- Was it
proven beyond a reasonable doubt?
- If they did and you find unanimously they did that, you
must follow the law and find them either guilty or not guilty of
that charge.'"
- You then apply the law to those facts as I
state it to you, and you must accept and follow 

In [22]:
# Calculer le score de silhouette pour évaluer la qualité du clustering
silhouette_avg = silhouette_score(sentence_embeddings, labels)
print(f"\nSilhouette Score: {silhouette_avg}")


Silhouette Score: 0.15657681226730347


# BM25

In [23]:
# Installer la bibliothèque si nécessaire
# %pip install rank-bm25

In [24]:
from rank_bm25 import BM25Okapi
from nltk.tokenize import word_tokenize

nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\elisa\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [25]:
# Exemple de documents (textes juridiques)
# documents = [
#     "a cat is a feline and likes to purr",
#     "a dog is the human's best friend and loves to play",
#     "a bird is a beautiful animal that can fly",
#     "a fish is a creature that lives in water and swims"
# ]

In [10]:
# Dictionnaire pour stocker les documents tokenisés par cluster
tokenized_clusters = {}

# Tokenisation des documents pour chaque cluster
for i, sentences in cluster.items():
    tokenized_clusters[i] = [word_tokenize(sentence.lower()) for sentence in sentences]

# Initialiser un modèle BM25 pour chaque cluster
bm25_models = {}
for i, tokenized_docs in tokenized_clusters.items():
    bm25_models[i] = BM25Okapi(tokenized_docs)

NameError: name 'cluster' is not defined

In [27]:
# Exemple de requête
query = "law and legal rights"
tokenized_query = word_tokenize(query.lower())

In [28]:
# Calcul des scores BM25 pour tous les clusters
for cluster_id, bm25 in bm25_models.items():
    # Calcul des scores pour la requête dans chaque cluster
    scores = bm25.get_scores(tokenized_query)
    # print(f"Scores pour la requête dans le cluster {cluster_id}: {scores}")

In [29]:
# for cluster_id, bm25 in bm25_models.items():
#     # Récupérer les documents les plus pertinents pour la requête
#     top_docs = bm25.get_top_n(tokenized_query, tokenized_clusters[cluster_id], n=1) # n=1 peut être ajusté pour obtenir plus de documents
#     print(f"\nCluster {cluster_id}:")
#     for doc in top_docs:
#         sentence = ' '.join(doc)
#         print(sentence)

best_sentences = []

for cluster_id, bm25 in bm25_models.items():
        # Récupérer les phrases les plus pertinentes pour la requête dans ce cluster
        top_docs = bm25.get_top_n(tokenized_query, tokenized_clusters[cluster_id], n=2)

        # Extraire les phrases pertinentes et les ajouter à la liste
        for doc in top_docs:
            sentence = ' '.join(doc)  # Convertir le tokenized doc en phrase
            best_sentences.append(sentence)

# Combiner les meilleures phrases en un paragraphe
paragraph_gen = ' '.join(best_sentences)
    
# Afficher le paragraphe généré
print("\nParagraphe généré à partir des meilleures phrases :")
print(paragraph_gen)


Paragraphe généré à partir des meilleures phrases :
for cert . to pet . accordingly , the comments made and not made by the court to the jury did not coerce a particular verdict or deny packer any constitutional rights . '' you then apply the law to those facts as i state it to you , and you must accept and follow the law . after the foreman indicated that further deliberations would be helpful , the judge gave the following instruction to the jury : < /p > < blockquote > '' 'what you do is-like i think what the instructions were-you apply the facts to the law and you arrive at a decision . this time she complained about `` 'feeling [ s ] of distrust and disrespect from the other jurors , ' '' and said that '' ' i have reached a point of anger , and i do n't believe i can be objective . ' '' you should read the full case before relying on it for legal research purposes. < /p > this site is protected by recaptcha and the google < a href= '' https : //policies.google.com/privacy '' targ

In [5]:
train[0]['raw_target']

paragraph_target = train[0]['raw_target']['facts_of_the_case'] + train[0]['raw_target']['question'] + train[0]['raw_target']['conclusion']

In [8]:
paragraph_target

'<p>William Packer was convicted in a California state court for second-degree murder and attempted murder. During jury deliberations, a juror requested dismissal on two separate occasions. In order to avoid having to restart deliberations with a replacement juror, the judge asked the juror to continue. After further conversations between the judge and the juror in question as well as the foreperson of the jury, the jury returned a guilty verdict. On appeal to the Court of Appeals of California, Packer argued that the state trial judge had coerced the jury’s verdict in violation of his Fourteenth Amendment right to due process. The appellate court affirmed his conviction and held that, as long as the judge only encouraged further deliberations and did not coerce a certain verdict, there was no violation of due process.</p>\n<p>Packer petitioned for habeas relief from the federal district court, which denied his petition. The US. Court of Appeals for the Ninth Circuit reversed the distr

In [32]:
score = fct.rouge_evaluations(paragraph_gen, paragraph_target)
print(score)

   Metric  Precision    Recall  F1-Score
0  rouge1   0.550725  0.046441  0.085658
1  rouge2   0.149091  0.012531  0.023118
2  rougeL   0.210145  0.017721  0.032685


M25 clean   

   Metric  Precision    Recall  F1-Score
0  rouge1   0.257246  0.467105  0.331776
1  rouge2   0.080000  0.145695  0.103286
2  rougeL   0.126812  0.230263  0.163551

M25 brut

   Metric  Precision    Recall  F1-Score
0  rouge1   0.449275  0.038871  0.071552
1  rouge2   0.109091  0.009407  0.017321
2  rougeL   0.184783  0.015987  0.029429

legal pegasus

   Metric  Precision    Recall  F1-Score
0  rouge1   0.919811  0.560345  0.696429
1  rouge2   0.791469  0.481268  0.598566
2  rougeL   0.768868  0.468391  0.582143


In [8]:
train_path = 'SCOTUS/train.json'
dev_path = 'SCOTUS/dev.json'

with open(train_path, 'r', encoding="utf-8") as f:
    train = json.load(f)

document = train[0]["raw_source"]

train[0]['raw_target']
paragraph_target = train[0]['raw_target']['facts_of_the_case'] + train[0]['raw_target']['question'] + train[0]['raw_target']['conclusion']

sentences = fct.sent_segmentation(document, method='nltk')
summary = fct.bb25LegalSum(sentences, "bert-base-uncased", 5)

bb25_rouge = fct.rouge_evaluations(" ".join(summary), paragraph_target)

- 01-1765.
- 01-1765.
- They do this or not do this?
- This element, this element, this
element?
- You
determine the facts.
- App.
- to Pet.
- for
Cert.
- H-15 to H-16 (citations omitted).
- See id., at 579, n. 10.
- App.
- to Pet.
- for Cert.
- H-15.
- to Pet.
- for
Cert.
- Awareness of this Court's
cases is not even required, so long as neither the reasoning nor
the result of the state-court decision contradicts them.
- Awareness of this Court's
cases is not even required, so long as neither the reasoning nor
the result of the state-court decision contradicts them.
- But they do
not have a right to not deliberate.
- They must deliberate and follow
the rules and laws as I state it to them.'"
- Was it
proven beyond a reasonable doubt?
- If they did and you find unanimously they did that, you
must follow the law and find them either guilty or not guilty of
that charge.'"
- You then apply the law to those facts as I
state it to you, and you must accept and follow the law.
- You can't
mak

In [9]:
bb25_rouge

Metric  Precision    Recall  F1-Score
0  rouge1   0.550725  0.046441  0.085658
1  rouge2   0.149091  0.012531  0.023118
2  rougeL   0.210145  0.017721  0.032685